### ISSUE TITLE
Сonfigurable factories

### ISSUE THREAD

**Issue by skar** [zendframework/zend-servicemanager#128]
Hi, all,

I have tourbles with tons of factories (controllers/actions/simple services/etc) and I think to do simple dependencies configuration like:

```php
'dep-name' => [
    'class' => MyService::class,
    'deps' => [
        Dep1::class,
        Dep2::class,
    ],
],
```
It's will be alternative for `factories`, `invokables` etc and works like a factory sample:

```php
'dep-name' => function($container) {
    return new MyService($container->get(Dep1::class), $container->get(Dep2::class));
},
```

In the future it can be used for annotations/reflection support maybe.

I want to implement it.

But I need help from you!
First, I need discussions and ideas for implementation (config examples, naming etc).
Second, I need support (moral).

**Comment by Ocramius** [comment]
As discussed on the mailing list, this is a good idea, but:

 * we need to enforce string keys for the parameters, as it is impossible to override them otherwise (merged config)
 * we need to type-check all parameters, and throw meaningful exceptions in case of problems
 * we may (further down the line) compile these definitions into factories
 * we need benchmarks for this, even if the first version will not actually be performing very well
 * we may look at symfony/dependency-injection to avoid writing our own incompatible DSL for no good reason

**Comment by skar** [comment]
> we need to enforce string keys for the parameters, as it is impossible to override them otherwise (merged config)

Stdlib\ArrayUtils::merge [can](https://github.com/zendframework/zend-stdlib/blob/master/src/ArrayUtils.php#L277) merge numeric keys without override, but not array_merge.
I think it will be the will of the end user, but it should be described in the documentation.

> we need to type-check all parameters, and throw meaningful exceptions in case of problems

Maybe typehint can help us?

> we need benchmarks for this, even if the first version will not actually be performing very well

Any examples?


> Avoid this bit: has been only pain so far ;-)

Keyword 'maybe' =)

**Comment by Koopzington** [comment]
type-check parameters?
If the constructor of the Service has typehinted it's parameters you'd automatically get meaningful errors when trying to pass instances to it that aren't what the constructors expect.
```php
final class Dep2
{
}

final class MyService
{
    public function __construct(Dep1 $dep1, Dep2 $dep2)
    {
    }
}

$myService = new MyService(new Dep2(), new Dep3());

Fatal error: Uncaught TypeError: Argument 1 passed to MyService::__construct() must be an instance of Dep1, instance of Dep2 given

Fatal error: Uncaught Error: Class 'Dep3' not found
```
Aren't those meaningful enough?

**Comment by GeeH** [comment]
This is a good idea.

**Comment by GeeH** [comment]
Didn't mean to enter then, I would suggest the dependencies should be service manager keys?

**Comment by skar** [comment]
Wait a minute, how about Zend Di? =)
As explained in the mailing list Zend SM - factories, Zend Di - configs.
And it's functionality must be in Zend Di.

**Comment by tasmaniski** [comment]
Great feature! This is exactly what will solve 90% factories in my projects. 

> I would suggest the dependencies should be service manager keys

Is there any reason not to work with other keys as 
_controllers_, _view_helpers_ or _controller_plugins_?

Although, feature doesn't need to be an integral part of the SM and abstract factory above is very good replacement, I would like to see this feature officially as part of the SM (and in it's documentation).

**Comment by GeeH** [comment]
@skar can you provide a link to mailing list discussion please?

**Comment by skar** [comment]
@GeeH http://zend-framework-community.634137.n4.nabble.com/Simple-zend-servicemanager-td4662889.html

And `invokables` was removed from SM (replaced by `InvokableFactory`).
After all, I think we must use Zend\Di for 'configurable factories'.

@Ocramius what do you think?

**Comment by GeeH** [comment]
> And invokables was removed from SM (replaced by InvokableFactory).

I don't understand what point this brings to the discussion.

> After all, I think we must use Zend\Di for 'configurable factories'.

Is Zend\Di still even a thing? I propose we drop Zend\Di and mark it as unmaintained even if it is not already. This is an optional AbstractFactory that can be enabled on a per-project basis, has zero overhead if not enabled and no perceivable performance difference if it is enabled. It depends on the existing ServiceManager (or at least a container-interop container) to find it's dependencies. For me, it's place is where it is, in the ServiceManager namespace.

**Comment by GeeH** [comment]
> controllers, view_helpers or controller_plugins?

Current implementation it only works for the service locator the abstract factory is added to.

**Comment by skar** [comment]
> I don't understand what point this brings to the discussion.
Even this config was replaced by factory.

AbstractFactories has unnecessary overhead for this task.

**Comment by tasmaniski** [comment]
> Is Zend\Di still even a thing? I propose we drop Zend\Di and mark it as unmaintained even if it is not already.  

I too have had the same thought. 

It is possible to register abstract factory for controllers also, 
only different in **getServiceLocator()**

`
$serviceLocator->getServiceLocator()->get('config'); 
`

Unfortunately for view_helper and controller_plugin I can not find a replacement.

For SM V3 regular factory should be fine (instead of abstract factory). 
There will be more things in config but only one factory for all services. 
That should bring some performance too.

**Comment by Koopzington** [comment]
Can't we extend the [ServiceManager](https://github.com/zendframework/zend-servicemanager/blob/master/src/ServiceManager.php#L307) to parse a new key like... idk 'factory_configs'?
Since all PluginManagers extend the ServiceManager you could then have those 'factory_configs' in 'service_manager', 'view_helpers', 'controller_plugins', etc.
If the config has that key then add an ConfigAbstractFactory to the PluginManager and pass the defined config to that ConfigAbstractFactory's constructor instead of letting it grab the merged config from the Container.

**Comment by weierophinney** [comment]
Resolved with #146, to be released soon with 3.2.0.
